### PR TITLE
GUIFontTTF: Fix array outof bound error in some corner cases

### PR DESCRIPTION
**Pull request by naseef**
## Description

This will fix array out of bound error in ascii character caching logic in `CGUIFontTTF`.  Changed the logic of caching ascii chars ( letter<255) with their letter value instead of the glyph index
## Motivation and context

see https://github.com/xbmc/xbmc/pull/19944#issuecomment-886221405
## How has this been tested?

Tested on Linux and windows build
## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by naseef** [comment]
@thexai for your review please

**Comment by thexai** [comment]
Thank you for submitting the fix

File: xbmc/guilib/GUIFontTTF.cpp [hunk 1]
```diff
@@ -805,7 +805,7 @@ CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr, FT_UInt glyph
   {
     if (m_char[i].letter < 255)
     {
-      character_t ch = ((m_char[i].glyphAndStyle & 0xffff0000) >> 8) | (m_char[i].glyphIndex);
+      character_t ch = ((m_char[i].glyphAndStyle & 0xffff0000) >> 8) | (m_char[i].letter);
```
Comment (thexai) [inline xbmc/guilib/GUIFontTTF.cpp:808]:
```suggestion
      character_t ch = ((m_char[i].glyphAndStyle & 0xffff0000) >> 8) | m_char[i].letter;
```
Parenthesis is not necessary here

**Review by thexai — APPROVED** [review]

**Comment by thexai** [comment]
Merged too fast :(
Still an issues with this: font used by Debug Info OSD uses "ligatures":

Captura de pantalla 2021-07-31 180719

See: https://typofonderie.com/font-support/opentype-features/

This is the reason for `glyphIndex`  = 33436 for letter 'f' because 'ff' , 'fi' and 'fl' sequences are formed with single glyph (probably others too). Then ASCII letter (< 255) does not guarantee glyphIndex < 2048.

Without this PR potential crash happen and with this PR bad font render occurs for determined letters combinations:

Sin título

Should be: buffers, flip, off

**Comment by thexai** [comment]
Return to use `glyphIndex` and check for `ch < LOOKUPTABLE_SIZE` may be the solution:

```c++
    if (m_char[i].letter < 255)
    {
      character_t ch = ((m_char[i].glyphAndStyle & 0xffff0000) >> 8) | m_char[i].glyphIndex;
      if (ch < LOOKUPTABLE_SIZE)
        m_charquick[ch] = m_char+i;
    }
```
This just means that the letter combinations ff, etc. they are not treated as ASCII characters and are not cached but are rendered correctly:

Sin título2

**Comment by naseef** [comment]
Oh !!!!
I think `ch < LOOKUPTABLE_SIZE` might cause issue if the font stores ascii letters on glymphs indexes greater than 256.

What you think about this solution. Keep `letter` as in this PR and check whether the actual rendering character is not a shaped character 
```
 if (letter < 255 && glyphIndex==FT_Get_Char_Index(m_face, m_char[i].letter))
  {
    character_t ch = (style << 8) | letter;
    if (ch < LOOKUPTABLE_SIZE && m_charquick[ch])
      return m_charquick[ch];
  }
```

and
```
    if (m_char[i].letter < 255 && m_char[i].glyphIndex==FT_Get_Char_Index(m_face, m_char[i].letter))
    {
      character_t ch = ((m_char[i].glyphAndStyle & 0xffff0000) >> 8) | m_char[i].letter;
      if (ch < LOOKUPTABLE_SIZE)
        m_charquick[ch] = m_char+i;
    }
```

**Comment by thexai** [comment]
Maybe @enen92 can help with this. This is not my area and I do not fully understand the functionality of the code. I can detect things that can cause problems but there are also other aspects to consider such as performance penalties (this code is performance critical because it runs continuously while the GUI is rendering).

Ping also @lrusak

**Comment by naseef** [comment]
I don't know the standard ways to benchmark the time period of a function,but I tried the snippet in [this answer](https://stackoverflow.com/a/27856440) to know how much time `FT_Get_Char_Index(m_face, letter)` takes to complete. Below is the output on my Intel i5 laptop.

```
elapsed time: 1.94e-07s
elapsed time: 1.9e-07s
elapsed time: 9.6e-08s
elapsed time: 1.7e-07s
elapsed time: 1.07e-07s
elapsed time: 2.94e-07s
elapsed time: 9.4e-08s
elapsed time: 9.4e-08s
elapsed time: 1.17e-07s
elapsed time: 1.17e-07s
elapsed time: 1.3e-07s
elapsed time: 1.37e-07s
elapsed time: 1.25e-07s
elapsed time: 1.18e-07s
elapsed time: 1.18e-07s
elapsed time: 1.3e-07s
elapsed time: 1.34e-07s
elapsed time: 1.27e-07s
elapsed time: 1.2e-07s
elapsed time: 1.61e-07s
elapsed time: 1.43e-07s
elapsed time: 1.26e-07s
elapsed time: 1.99e-07s
elapsed time: 8.4e-08s
elapsed time: 1.1e-07s
elapsed time: 9.5e-08s
elapsed time: 1.3e-07s
elapsed time: 1.95e-07s
elapsed time: 1.07e-07s
elapsed time: 1.07e-07s
elapsed time: 1.15e-07s
elapsed time: 1.53e-07s
elapsed time: 1.99e-07s
elapsed time: 1.2e-07s
elapsed time: 1.01e-07s
elapsed time: 1.21e-07s
elapsed time: 1.89e-07s
elapsed time: 3.91e-07s
elapsed time: 1.32e-07s
elapsed time: 1.19e-07s
elapsed time: 1.06e-07s
elapsed time: 1.01e-07s
elapsed time: 1.14e-07s
elapsed time: 2.1e-07s
elapsed time: 1.28e-07s
elapsed time: 1.11e-07s
elapsed time: 1.12e-07s
elapsed time: 1.34e-07s
elapsed time: 1.19e-07s
elapsed time: 1.22e-07s
elapsed time: 1.35e-07s
elapsed time: 1.46e-07s
elapsed time: 4.84e-07s
elapsed time: 1.3e-07s
elapsed time: 1.42e-07s
elapsed time: 1.33e-07s
elapsed time: 1.26e-07s
elapsed time: 1.33e-07s
elapsed time: 1.02e-07s
elapsed time: 1.04e-07s
elapsed time: 1.37e-07s
elapsed time: 1.76e-07s
elapsed time: 1.45e-07s
elapsed time: 2.7e-07s
elapsed time: 1.47e-07s
elapsed time: 1.35e-07s
elapsed time: 1.12e-07s
elapsed time: 1.16e-07s
elapsed time: 1.15e-07s
elapsed time: 1.01e-07s
elapsed time: 1.15e-07s
elapsed time: 1.49e-07s
elapsed time: 1.6e-07s
elapsed time: 1.21e-07s
elapsed time: 1.47e-07s
elapsed time: 1.3e-07s
elapsed time: 1.17e-07s
elapsed time: 1e-07s
elapsed time: 1.22e-07s
elapsed time: 1.02e-07s
elapsed time: 1.09e-07s
elapsed time: 3.78e-07s
elapsed time: 1.08e-07s
elapsed time: 1.23e-07s
elapsed time: 1.44e-07s
elapsed time: 1.26e-07s
elapsed time: 1.14e-07s
elapsed time: 1.15e-07s
elapsed time: 1.16e-07s
elapsed time: 1.15e-07s
elapsed time: 1.07e-07s
elapsed time: 1.45e-07s
elapsed time: 1.19e-07s
elapsed time: 1.33e-07s
elapsed time: 1.25e-07s
elapsed time: 1.16e-07s
elapsed time: 1.05e-07s
elapsed time: 1.18e-07s
elapsed time: 1.24e-07s
elapsed time: 1.75e-07s
elapsed time: 1.13e-07s
elapsed time: 1.24e-07s
elapsed time: 1.1e-07s
elapsed time: 1.07e-07s
elapsed time: 1e-07s
elapsed time: 8.7e-08s
elapsed time: 1.3e-07s
elapsed time: 1.53e-07s
elapsed time: 1.24e-07s
elapsed time: 1.28e-07s
elapsed time: 1.08e-07s
elapsed time: 1.13e-07s
elapsed time: 1.08e-07s
elapsed time: 1.12e-07s
elapsed time: 1.27e-07s
elapsed time: 1.18e-07s
elapsed time: 1.15e-07s
elapsed time: 3.7e-07s
elapsed time: 1.13e-07s
elapsed time: 1.22e-07s
elapsed time: 1.08e-07s
elapsed time: 3.78e-07s
elapsed time: 1.17e-07s
elapsed time: 1.32e-07s
elapsed time: 1.17e-07s
elapsed time: 3.73e-07s
elapsed time: 1.07e-07s
elapsed time: 1.6e-07s
elapsed time: 1.2e-07s
elapsed time: 1.11e-07s
elapsed time: 1.56e-07s
elapsed time: 1.17e-07s
elapsed time: 1.18e-07s
elapsed time: 1.14e-07s
elapsed time: 1.21e-07s
elapsed time: 3.82e-07s
elapsed time: 1e-07s
elapsed time: 1.08e-07s
elapsed time: 1.21e-07s
elapsed time: 1.21e-07s
elapsed time: 1.21e-07s
elapsed time: 1.29e-07s
elapsed time: 1.1e-07s
elapsed time: 1e-07s
elapsed time: 1.17e-07s
elapsed time: 1.18e-07s
elapsed time: 1.13e-07s
elapsed time: 1.38e-07s
elapsed time: 3.86e-07s
elapsed time: 1.38e-07s
elapsed time: 1.22e-07s
elapsed time: 1.13e-07s
elapsed time: 1.25e-07s
elapsed time: 1.24e-07s
elapsed time: 1.28e-07s
elapsed time: 9.9e-08s
elapsed time: 2.307e-06s
elapsed time: 2.327e-06s
elapsed time: 9.53e-07s
elapsed time: 7.51e-07s
elapsed time: 4.64e-07s
elapsed time: 5.59e-07s
elapsed time: 4.32e-07s
elapsed time: 2.66e-07s
elapsed time: 7.2e-08s
elapsed time: 3.49e-07s
elapsed time: 1.91e-07s
elapsed time: 7.2e-08s
elapsed time: 2.31e-07s
elapsed time: 6.5e-08s
elapsed time: 7.1e-08s
elapsed time: 7.1e-08s
elapsed time: 7.1e-08s
elapsed time: 9.1e-07s
elapsed time: 2.03e-07s
elapsed time: 4.11e-07s
elapsed time: 3.33e-07s
elapsed time: 9.13e-07s
elapsed time: 5.46e-07s
elapsed time: 2.45e-07s
elapsed time: 4.09e-07s
elapsed time: 4.11e-07s
elapsed time: 1.21e-07s
elapsed time: 1.58e-07s
elapsed time: 1.86e-07s
elapsed time: 5.99e-07s
elapsed time: 4.39e-07s
elapsed time: 4.92e-07s
elapsed time: 8.1e-07s
elapsed time: 1.17e-07s
elapsed time: 2.11e-07s
elapsed time: 5.05e-07s
elapsed time: 1.05e-07s
elapsed time: 3.38e-07s
elapsed time: 1.97e-07s
elapsed time: 1.17e-07s
elapsed time: 1.12e-07s
elapsed time: 1.99e-07s

```

**Comment by thexai** [comment]
It's not only time takes `FT_Get_Char_Index` also these very common letters sequences `ff`, `fl` `fi` are not cached then. The desirable behavior with ASCII characters is all are cached and `CGUIFontTTF::GetCharacter`  returns quick here: https://github.com/xbmc/xbmc/blob/5f726439352831c17486a77ac1a6325b7f778f56/xbmc/guilib/GUIFontTTF.cpp#L742

This means that if all the characters on the screen are ASCII and the text does not change significantly, it is used lookup table all the time and never are used these loop blocks (until text changes):

https://github.com/xbmc/xbmc/blob/5f726439352831c17486a77ac1a6325b7f778f56/xbmc/guilib/GUIFontTTF.cpp#L750-L759

and 

https://github.com/xbmc/xbmc/blob/5f726439352831c17486a77ac1a6325b7f778f56/xbmc/guilib/GUIFontTTF.cpp#L803-L811

If a single character is not in the lookup table then this is not possible and loops are executed "all the time" even if text rendered not changes. One example is Debug Info OSD as contains ff and Debug Info Video OSD (Alt+O) contains both ff and fl.

These loops are executed for every text letter and all are repeated for every frame rendered. Probably with a text of only 100 letters at 60 fps are repeated 6000 times in one second. That is the reason why it requires optimization to the maximum.

I think I have come up with a way to do what I want in a simple way and later I will upload a PR for it.

**Pull request merged**